python/dgl/graphbolt/itemset.py

```diff
@@ -359,65 +326,79 @@ class ItemSetDict:
-            ]
-            self._offsets = torch.tensor(offsets).cumsum(0)
-        except TypeError:
-            self._offsets = None
-
-    def __iter__(self) -> Iterator:
-        for key, itemset in self._itemsets.items():
-            for item in itemset:
-                yield {key: item}
+        offset = [0] + [len(itemset) for itemset in self._itemsets.values()]
+        self._offsets = torch.tensor(offset).cumsum(0)
+        self._length = int(self._offsets[-1])
+        self._keys = list(self._itemsets.keys())
 
     def __len__(self) -> int:
-        return sum(len(itemset) for itemset in self._itemsets.values())
+        return self._length
 
-    def __getitem__(self, idx: Union[int, slice]) -> Dict[str, Tuple]:
-        if self._offsets is None:
-            raise TypeError(
-                f"{type(self).__name__} instance doesn't support indexing."
-            )
-        total_num = self._offsets[-1]
-        if isinstance(idx, int):
-            if idx < 0:
-                idx += total_num
-            if idx < 0 or idx >= total_num:
+    def __getitem__(self, index: Union[int, slice, Iterable[int]]):
+        if isinstance(index, int):
+            if index < 0:
+                index += self._length
+            if index < 0 or index >= self._length:
                 raise IndexError(f"{type(self).__name__} index out of range.")
-            offset_idx = torch.searchsorted(self._offsets, idx, right=True)
+            offset_idx = torch.searchsorted(self._offsets, index, right=True)
             offset_idx -= 1
-            idx -= self._offsets[offset_idx]
-            key = list(self._itemsets.keys())[offset_idx]
-            return {key: self._itemsets[key][idx]}
-        elif isinstance(idx, slice):
-            start, stop, step = idx.indices(total_num)
-            assert step == 1, "Step must be 1."
+            index -= self._offsets[offset_idx]
+            key = self._keys[offset_idx]
+            return {key: self._itemsets[key][index]}
+        elif isinstance(index, slice):
+            start, stop, step = index.indices(self._length)
+            if step != 1:
+                return self.__getitem__(list(range(start, stop, step)))
             assert start < stop, "Start must be smaller than stop."
             data = {}
             offset_idx_start = max(
                 1, torch.searchsorted(self._offsets, start, right=False)
             )
-            keys = list(self._itemsets.keys())
             for offset_idx in range(offset_idx_start, len(self._offsets)):
-                key = keys[offset_idx - 1]
+                key = self._keys[offset_idx - 1]
                 data[key] = self._itemsets[key][
                     max(0, start - self._offsets[offset_idx - 1]) : stop
                     - self._offsets[offset_idx - 1]
                 ]
                 if stop <= self._offsets[offset_idx]:
                     break
             return data
+        elif isinstance(index, Iterable):
```

Comment:
This is new feature. Please make sure it's well tested.

Reply:
Do we have any requirement for this feature? If not necessary, I recommend not to offer it. More complicated incurs more burden on maintenance.

Reply:
any specific testcase for this feature?

Reply:
This feature is essential for 'real' sharding between distributed replicas.

Reply:
As discussed offline, please do comprehensive perf test on ItemSampler which leverages this access logic. It may have perf issue.